docs/developer/snmp.rst

```diff
@@ -0,0 +1,88 @@
+twisted as an I/O framework, which integrates nicely with OCS/SOCS. SOCS makes this
+twisted interface for SNMP available via the SNMPTwister class.
+
+.. _pysnmp: http://snmplabs.com/pysnmp/contents.html
+
+MIB to Python Conversion
+------------------------
+For developers adding a new SNMP monitoring OCS Agent, you may need to convert
+a MIB file to python. This can be done with mibdump.py, a conversion script
+provided by pysmi. Other useful (linux) packages for debugging include smitools
+and snmp.
+
+An example which converted the MBG-SNMP-LTNG-MIB .mib file::
+
+    $ mibdump.py --mib-source . --mib-source /usr/share/snmp/mibs/ MBG-SNMP-LTNG-MIB
+    Source MIB repositories: ., /usr/share/snmp/mibs/
+    Borrow missing/failed MIBs from: http://mibs.snmplabs.com/pysnmp/notexts/@mib@
+    Existing/compiled MIB locations: pysnmp.smi.mibs, pysnmp_mibs
+    Compiled MIBs destination directory: /home/bjk49/.pysnmp/mibs
+    MIBs excluded from code generation: INET-ADDRESS-MIB, PYSNMP-USM-MIB, RFC-1212, RFC-1215, RFC1065-SMI, RFC1155-SMI, RFC1158-MIB, RFC1213-MIB, SNMP-FRAMEWORK-MIB, SNMP-TARGET-MIB, SNMPv2-CONF, SNMPv2-SMI, SNMPv2-TC, SNMPv2-TM, TRANSPORT-ADDRESS-MIB
+    MIBs to compile: MBG-SNMP-LTNG-MIB
+    Destination format: pysnmp
+    Parser grammar cache directory: not used
+    Also compile all relevant MIBs: yes 
+    Rebuild MIBs regardless of age: no
+    Dry run mode: no
+    Create/update MIBs: yes 
+    Byte-compile Python modules: yes (optimization level no) 
+    Ignore compilation errors: no
+    Generate OID->MIB index: no
+    Generate texts in MIBs: no
+    Keep original texts layout: no
+    Try various file names while searching for MIB module: yes 
+    Created/updated MIBs: MBG-SNMP-LTNG-MIB, MBG-SNMP-ROOT-MIB, SNMPv2-MIB
+    Pre-compiled MIBs borrowed: 
+    Up to date MIBs: SNMPv2-CONF, SNMPv2-SMI, SNMPv2-TC
+    Missing source MIBs: 
+    Ignored MIBs: 
+    Failed MIBs: 
+
+.. note::
+    There can be several "gotchas" during the conversion process, mostly to do
+    with the naming of the .mib files and location of where they can be found. This
+    example specifies the source locations manually. The ``--debug all`` flag can
+    be useful in debugging conversion problems.
+
+Examples
+--------
+
+.. note::
+
+    These examples assume you're in the reactor context, i.e. running in an OCS
+    Agent non-blocking task or process. This primarily means we don't start the
+    reactor or any sort of runner ourselves.
+
+Initializing an SNMPTwister::
+
+    from socs.snmp import SNMPTwister
+
+    snmp = SNMPTwister(address, port)
+
+    get_list = [ObjectType(ObjectIdentity('MBG-SNMP-LTNG-MIB', 'mbgLtNgRefclockState', 1)),
+                ObjectType(ObjectIdentity('MBG-SNMP-LTNG-MIB', 'mbgLtNgSysPsStatus', 1)),
+                ObjectType(ObjectIdentity('MBG-SNMP-LTNG-MIB', 'mbgLtNgSysPsStatus', 2))]
+
+    result = yield snmp.get(get_list)
+
```

Comment:
I think this example is a little too sketchy.  Consider fleshing out to a full working script -- would only ~double the number of lines.  Need some imports (where is ObjectType from?) and a ```@inlineCallbacks``` function and a reactor run.  Would still be helpful (the pysnmp examples are pretty intimidating).  Is it possible to run without an MIB?  Such a bare bones +twisted starting point could be really useful for people working on new Agents.

Reply:
I would love to make this a standalone script. I do not know how to get this to work though. Would be happy to discuss it on Slack sometime.

ObjectType is no longer needed (it's from pysnmp), since SNMPTwister now accepts tuples per your suggestion above. This example is extremely specific to the M1000, so no, it's not possible to run this without the MIB. Some demo SNMP target could probably be made though...

Reply:
Alright, in typical fashion I have figured it out immediately after saying otherwise. Will update in a moment...

Reply:
So the example I've come up with that works looks like this:
```
    from twisted.internet import reactor
    from socs.snmp import SNMPTwister
    
    # Setup communication with M1000
    snmp = SNMPTwister('10.10.10.186', 161)
    
    # Define OIDs to query
    get_list = [('MBG-SNMP-LTNG-MIB', 'mbgLtNgRefclockState', 1),
                ('MBG-SNMP-LTNG-MIB', 'mbgLtNgSysPsStatus', 1),
                ('MBG-SNMP-LTNG-MIB', 'mbgLtNgSysPsStatus', 2)]
    
    def printResult(x):
        print(x)
    
    # Run the twisted SNMP GET and print the results
    d = snmp.get(get_list)
    d.addCallback(printResult)
    
    # Shut things down by asking the reactor to stop after 3 seconds
    reactor.callLater(3, reactor.stop)
    reactor.run()
```

My problem with this is that it doesn't look like what you'll end up writing from within an agent, which is what I was trying to go for with the other example, as I don't expect people to be using SNMPTwister much outside of Agents, at least, that's outside the scope of this guide.

Reply:
How about:
```
@inlineCallbacks
def query_snmp():
    x = yield snmp.get(get_list)
    print(x)
    reactor.stop()
```

Reply:
Mhm, I had tried something like this, it hangs and doesn't print anything.

Reply:
Debugged offline, example updated.